CRM/Utils/REST.php

```diff
@@ -636,4 +627,35 @@ public function loadCMSBootstrap() {
     }
   }
 
+  /**
+   * Does this request appear to be a web-service request?
+   *
+   * This is used to mitigate CSRF risks.
+   *
+   * @return bool
+   *   TRUE if the current request appears to either XMLHttpRequest or non-browser-based.
+   *       Indicated by either (a) custom headers like `X-Request-With`/`X-Civi-Auth`
+   *       or (b) strong-secret-params that could theoretically appear in URL bar but which
+   *       cannot be meaningfully forged for CSRF purposes (like `?api_key=SECRET` or `?_authx=SECRET`).
+   *   FALSE if the current request looks like a standard browser request. This request may be generated by
+   *       <A HREF>, <IFRAME>, <IMG>, `Location:`, or similar CSRF vector.
+   */
+  protected static function isWebServiceRequest(): bool {
+    if (($_SERVER['HTTP_X_REQUESTED_WITH'] ?? NULL) === 'XMLHttpRequest') {
+      return TRUE;
+    }
+
+    $authx = \CRM_Core_Session::singleton()->get('authx');
```

Comment:
This needs an extra guard - the `authx` flows are only suitable alternatives if the `authx` extension is actually enabled.

Reply:
Strike that, it's OK. The only way the session-variable exists is if `authx` is enabled and has approved the credential.